pkg/download/tftp.go

```diff
@@ -0,0 +1,66 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package download
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/pin/tftp"
+)
+
+// NewTFTPTransport returns an http.RoundTripper capable of handling the TFTP
+// protocol.
+func NewTFTPTransport() http.RoundTripper {
+	return &tftpRoundTripper{}
+}
+
+var _ http.RoundTripper = &tftpRoundTripper{}
+
+type tftpRoundTripper struct{}
+
+func (t *tftpRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	addr := req.URL.Host
+
+	if req.URL.Port() == "" {
+		addr += ":69"
+	}
+
+	c, err := tftp.NewClient(addr)
+	if err != nil {
+		return nil, err
+	}
+
+	w, err := c.Receive(req.URL.Path, "octet")
+	if err != nil {
+		return nil, err
+	}
+
+	buf := &bytes.Buffer{}
+
+	written, err := w.WriteTo(buf)
+	if err != nil {
+		return nil, err
+	}
+
+	if expected, ok := w.(tftp.IncomingTransfer).Size(); ok {
+		if written != expected {
+			return nil, fmt.Errorf("expected %d bytes, got %d", expected, written)
+		}
+	}
+
+	return &http.Response{
+		Status:        "200 OK",
+		StatusCode:    200,
+		Proto:         "TFTP/1.0",
+		ProtoMajor:    1,
+		ProtoMinor:    0,
+		Body:          ioutil.NopCloser(buf),
```

Comment:
@smira This is the only thing I wasn't too sure about, but as far as I can tell, the TFTP library doesn't return anything I need to close. Gonna merge, can follow up if you think this isn't good.

Reply:
I haven't looked at tftp library closely, but you return the buffer, so you can close tftp client by the time request is done

Reply:
and as it's all UDP, I believe there's nothing to close really

Reply:
yep... ok so I think this is good